PERF: Eager load Theme associations in Stylesheet Manager.

tgxworld opened the pull request:
Before this change, calling `StyleSheet::Manager.stylesheet_details`
for the first time resulted in multiple queries to the database. This is
because the code was modelled in a way where each `Theme` was loaded
from the database one at a time.

This PR restructures the code such that it allows us to load all the
theme records in a single query. It also allows us to eager load the
required associations upfront. In order to achieve this, I removed the
support of loading multiple themes per request. It was initially added
to support user selectable theme components but the feature was never
completed and abandoned because it wasn't a feature that we thought was
worth building.

tgxworld commented on app/helpers/application_helper.rb:
```diff
@@ -408,14 +408,19 @@ def topic_featured_link_domain(link)
 
+  def stylesheet_manager
+    return @stylesheet_manager if defined?(@stylesheet_manager)
+    @stylesheet_manager = Stylesheet::Manager.new(theme_id: theme_id)
```
Each request will work with an instance of `Stylesheet::Manager` which stores state of all the themes that has been loaded within a request. This allows us to reuse the state across method calls which wasn't possible previously.

tgxworld commented on app/models/theme.rb:
```diff
@@ -188,25 +201,28 @@ def self.clear_default!
     expire_site_cache!
   end
 
-  def self.transform_ids(ids, extend: true)
```
Drop support for `extend` here. It increases the matrix for caching leading to extra requests.

tgxworld commented on app/models/theme.rb:
```diff
@@ -188,25 +201,28 @@ def self.clear_default!
-
-      all_ids = [parent, *components]
+      all_ids =
+        if self.is_parent_theme?(parent)
```
No need to call `components_for` is theme is not a parent theme.

tgxworld commented on spec/requests/safe_mode_controller_spec.rb:
```diff
@@ -11,6 +11,8 @@
 
       get '/safe-mode'
+
+      expect(response.status).to eq(200)
```
This test has been 500ing leading to a false positive.

tgxworld commented on app/models/theme.rb:
```diff
@@ -188,25 +201,22 @@ def self.clear_default!
     expire_site_cache!
   end
 
-  def self.transform_ids(ids, extend: true)
-    return [] if ids.nil?
-    get_set_cache "#{extend ? "extended_" : ""}transformed_ids_#{ids.join("_")}" do
-      next [] if ids.blank?
-
-      ids = ids.dup
-      ids.uniq!
-      parent = ids.shift
+  def self.transform_ids(id)
```
Drop support for querying for multiple themes. The previous behaviour was too confusing as it assumes the first element in the array is the parent theme id while the other elements in the array are the components theme ids of the parent.

davidtaylorhq replied:
I think it might be good to add a validation here to make sure there isn't anywhere still passing an array. I took this PR for a spin locally, and it looks like theme modifiers are broken because of this issue (`ThemeModifierHelper` and `Theme.lookup_modifier` handle arrays).

Easy way to try this out is to install https://meta.discourse.org/t/topic-list-excerpts-theme-component/151520 locally as a component.

tgxworld commented on lib/stylesheet/manager/builder.rb:
```diff
@@ -0,0 +1,274 @@
+# frozen_string_literal: true
+
+class Stylesheet::Manager::Builder
```
This used to be `Stylesheet::Manager` but I need an instance of `Stylesheet::Manager` per request so moving this to another class as a builder. Copied and pasted instance methods from `Stylesheet::Manager` with some slight tweaks.

tgxworld commented on lib/stylesheet/manager.rb:
```diff
@@ -232,245 +131,165 @@ def self.cache_fullpath
-  end
+    Theme
+      .where(id: to_load_theme_ids)
+      .includes(:yaml_theme_fields, :theme_settings, :upload_fields, :builder_theme_fields)
```
This is where all of the savings happen.

tgxworld commented:
Test failure is a front end related.

OsamaSayegh reviewed (commented):
I've left a few minor points/comments, but overall this looks great to me 👌

davidtaylorhq commented on app/models/theme.rb:
```diff
@@ -335,14 +344,18 @@ def self.notify_theme_change(theme_ids, with_scheme: false, clear_manager_cache:
   def self.refresh_message_for_targets(targets, theme_ids)
-    targets.map do |target|
-      Stylesheet::Manager.stylesheet_data(target.to_sym, theme_ids)
+    theme_ids = [theme_ids] unless theme_ids === Array
```
Personal preference, but I prefer to use `theme_ids.is_a? Array` - this reads much nicer to me

davidtaylorhq commented on app/views/layouts/application.html.erb:
```diff
@@ -5,7 +5,7 @@
     <title><%= content_for?(:title) ? yield(:title) : SiteSetting.title %></title>
     <meta name="description" content="<%= @description_meta || SiteSetting.site_description %>">
-    <meta name="discourse_theme_ids" content="<%= theme_ids&.join(",") %>">
+    <meta name="discourse_theme_ids" content="<%= theme_id %>">
```
Should we update the `name` of this meta tag? (and all the JS which references it?)

tgxworld replied:
Good catch :+1:

davidtaylorhq reviewed (commented):
Took this for a spin locally, seems to work well overall. A couple of issues I noticed:

1. Theme modifiers are not working for components (see specific comment above)
2. Previewing a component which has no theme CSS is completely broken. (e.g. install [this component](https://meta.discourse.org/t/topic-list-excerpts-theme-component/151520?u=david) locally, click preview, everything breaks. If you add some simple CSS like `body{color: red}`, then things start working again)

tgxworld commented on app/views/layouts/no_ember.html.erb:
```diff
@@ -13,8 +13,11 @@
-  <%= build_plugin_html 'server:header' %>
+  <%- unless customization_disabled? %>
+    <%= theme_lookup("header") %>
+    <%= build_plugin_html 'server:header' %>
```
Interesting, I was actually just copy and pasting the logic from `application.html.erb` and it seems like we have `customization_disabled?` and `allow_plugins` helper methods. I've split the rendering of the plugin html into another conditional.

tgxworld commented on app/views/common/_discourse_publish_stylesheet.html.erb:
```diff
@@ -10,6 +10,6 @@
   <%= discourse_stylesheet_link_tag(file) %>
```
Good catch! I've fixed other spots in the code base too.

tgxworld commented:
> 1. Theme modifiers are not working for components (see specific comment above)

I've fixed most of the problems and added test cases as well :) 

> 2\. Previewing a component which has no theme CSS is completely broken. (e.g. install [this component](https://meta.discourse.org/t/topic-list-excerpts-theme-component/151520?u=david) locally, click preview, everything breaks. If you add some simple CSS like `body{color: red}`, then things start working again)

Interesting... it seems like previewing a theme component is totally broken even on master. I'll debug and submit a PR separately.

tgxworld commented on spec/lib/content_security_policy_spec.rb:
```diff
@@ -303,13 +303,23 @@ def theme_policy
       theme.theme_modifier_set.csp_extensions = ["script-src: https://from-theme-flag.script", "worker-src: from-theme-flag.worker"]
       theme.save!
 
+      child_theme = Fabricate(:theme, component: true)
+      theme.add_relative_theme!(:child, child_theme)
+      child_theme.theme_modifier_set.csp_extensions = ["script-src: https://child-theme-flag.script", "worker-src: child-theme-flag.worker"]
+      child_theme.save!
+
       expect(parse(theme_policy)['script-src']).to include('https://from-theme-flag.script')
+      expect(parse(theme_policy)['script-src']).to include('https://child-theme-flag.script')
       expect(parse(theme_policy)['worker-src']).to include('from-theme-flag.worker')
+      expect(parse(theme_policy)['worker-src']).to include('child-theme-flag.worker')
```
@davidtaylorhq I've improved our test suite here which didn't catch csp_extensions that are set in a child theme.

tgxworld commented on spec/requests/safe_mode_controller_spec.rb:
```diff
@@ -10,8 +10,18 @@
+
+      expect(response.status).to eq(200)
       expect(response.body).not_to include("My Custom Header")
+      expect(response.body).not_to include("data-theme-id=\"#{theme.id}\"")
```
@OsamaSayegh I've added a spec to ensure that themes can be disabled in safe mode.

tgxworld commented on spec/serializers/listable_topic_serializer_spec.rb:
```diff
@@ -0,0 +1,41 @@
+  fab!(:topic) { Fabricate(:topic) }
+
+  describe '#excerpt' do
+    it 'can be extended by theme modifiers' do
+      payload = TopicListItemSerializer.new(topic,
+        scope: Guardian.new,
+        root: false
+      ).as_json
+
+      expect(payload[:excerpt]).to eq(nil)
+
+      theme = Fabricate(:theme)
+
+      child_theme = Fabricate(:theme, component: true).tap do |t|
+        theme.add_relative_theme!(:child, t)
+      end
+
+      child_theme.theme_modifier_set.serialize_topic_excerpts = true
+      child_theme.save!
+
+      request = ActionController::TestRequest.new(
+        { resolved_theme_id: theme.id },
+        nil,
+        nil
+      )
+
+      guardian = Guardian.new(nil, request)
+
+      payload = TopicListItemSerializer.new(topic,
+        scope: guardian,
+        root: false
+      ).as_json
+
+      expect(payload[:excerpt]).to eq(topic.excerpt)
```
I've added a spec here for testing theme modifier set for `serialize_topic_excerpts` too.

OsamaSayegh reviewed (approved):
Looks great to me! Nice work!

tgxworld commented:
Thank you for the review @OsamaSayegh @davidtaylorhq
